### ISSUE TITLE
Any way to listen for websocket PINGs?

### ISSUE THREAD

**Issue by vincentjames501** [ztellman/aleph#332]
In node.js or the browser (and many other websocket clients/servers), you can listen for PING/PONG events as well as issue PINGs at any time manually. This functionality would be nice to have in both the client and server as we need these heartbeats in our system.

**Comment by ztellman** [comment]
You can do this by hooking into the underlying Netty ChannelPipeline, but I agree that having a more first-class mechanism for this would be valuable.  I'll leave this issue open while I consider what a decent approach to this would be.

**Comment by vincentjames501** [comment]
Thanks @ztellman . If anyone else runs into this and wants a work around, this is what I'm doing in the meantime.

```clojure
(let [^Channel ch (.ch req)
        ^ChannelPipeline cp (.pipeline ch)]
    (.addBefore cp
                "websocket-handler"
                "websocket-ping-handler"
                (netty/channel-handler
                  :channel-read
                  ([_ ctx msg]
                    (when (instance? WebSocketFrame msg)
                      (let [^WebSocketFrame msg msg]
                        (when (instance? PingWebSocketFrame msg)
                          ;; DO THINGS
                          )))
                    (.fireChannelRead ctx msg)))))
```

**Comment by vincentjames501** [comment]
I'm having a slight performance issue that I'll dig more into tomorrow. When running the same load test tool (it creates 10k websocket connections, randomly sends data to RabbitMQ, reads data back from RabbitMQ, and sends it to the client). The "responses" begin to take longer and longer with Aleph to eventually more than 60 seconds to get a response.

Here is my Aleph code in a nutshell (the same results can be had by not going through RabbitMQ). Often, if I keep letting it run I eventually run into this: https://github.com/netty/netty/issues/5304

```clojure
(defn handle-device-ws-req
  [req]
  (-> (d/let-flow [socket (http/websocket-connection req)]
       (let [from-device-msg-chan (async/chan)
             to-device-msg-chan (rabbit-mq/register! from-device-msg-chan (get-in req [:query-params "name"]))]
         (async/go-loop []
           (when-some [msg-to-send (async/<! to-device-msg-chan)]
             (s/put! socket msg-to-send)
             (recur)))
         (s/connect socket from-device-msg-chan)
         (s/on-closed socket #(async/close! from-device-msg-chan))))
      (d/catch
       (constantly non-websocket-request))))

(http/start-server (params/wrap-params handler)
                                 {:port (get-in config [:http :port])})
```

Here is the implementation with http-kit which keeps linear "response" times.

```clojure
(defn handle-device-ws-req
  [req]
  (http/with-channel req c
   (if (http/websocket? c)
     (let [from-device-msg-chan (async/chan)
           to-device-msg-chan (rabbit-mq/register! from-device-msg-chan (get-in req [:query-params "name"]))]
       (async/go-loop []
         (when-some [msg-to-send (async/<! to-device-msg-chan)]
           (http/send! c msg-to-send false)
           (recur)))
       (http/on-close c (fn [_]
                          (async/close! from-device-msg-chan)))
       (http/on-receive c (fn [data]
                            (async/go
                              (async/>! from-device-msg-chan data)))))
     non-websocket-request)))

(http/run-server (params/wrap-params handler)
                               {:port (get-in config [:http :port])})
```

Any thoughts @ztellman ? Anything I should check first?

**Comment by ztellman** [comment]
My guess is you're doing something blocking where you shouldn't, though offhand it's not clear to me where.  If you can provide some stack dumps via `jstack` when the issue crops up, that might help.

**Comment by vspinu** [comment]
Probably a duplicate of #322

**Comment by kachayev** [comment]
Closed with #364.
